docs/RFCS/leader_evaluated_raft.md

```diff
@@ -0,0 +1,425 @@
+- Feature Name: leader_evaluated_raft
+- Status: draft
+- Start Date: 2016-04-19
+- Authors: Tobias Schottdorf
+- RFC PR: #6166
+- Cockroach Issue: (one or more # from the issue tracker)
+
+
+# Summary
+
+Pursue an idea by @spencerkimball brought up in the context of
+[#5985](https://github.com/cockroachdb/cockroach/pull/5985): move the bulk of
+the logic involved in applying commands to a Replica upstream of Raft.
+
+# Motivation
+
+The [first exploration of
+migrations](https://github.com/cockroachdb/cockroach/pull/5985) highlighted
+the complexity concomitant with migrations on top of the current Raft execution
+model.
+
+As of writing, the leading replica is the only replica that proposes commands
+to Raft. However, each replica applies commands individually, which leads to
+large swaths of code living "downstream" of Raft. Since all of this code needs
+to produce identical output even during migrations, a satisfactory migration
+story is almost inachievable, a comment made by several reviewers (and the
+author).
+
+By having the proposing replica compute the effects on the key-value state
+before proposing the command to Raft, much of this downstream code moves
+upstream of Raft; naively what's left are simple writes to key-value pairs
+(i.e. all commands move upstream except for a simplified MVCCPut operation).
+While some complexity remains, this has promise for a much facilitated
+migration story, as well as de-triplicating work previously done by each
+individual replica.
+
+# Detailed design
+
+## Execution path
+
+### Proposing commands
+
+The main change in the execution path will be in `proposeRaftCommand`, which
+currently takes a `BatchRequest` and creates a `pendingCmd`, which currently
+has the form
+
+```go
+type pendingCmd struct {
+  ctx     context.Context
+  idKey   storagebase.CmdIDKey
+  done    chan roachpb.ResponseWithError // Used to signal waiting RPC handler
+  raftCmd struct {
+      RangeID       RangeID
+      OriginReplica ReplicaDescriptor
+      Cmd           BatchRequest
+  }
+}
+```
+
+and within which the main change is changing the type of `Cmd` from
+`BatchRequest` to a new type which carries the effects of the application of
+the batch, as computed (by the leader) earlier in `proposeRaftCommand`.
+
+For simplicity, in this draft we will assume that the `raftCmd` struct will
+be modified according to these requirements.
+
+The "effects" contain
+
+* any key-value writes: For example the outcome of a `Put`, `DeleteRange`, `CPut`, `BeginTransaction`, or `LeaderLease`.
+  Some alternatives exist: writes could be encoded as `[]MVCCKeyValue`, or even
+  as post-MVCC raw (i.e. opaque) key-value pairs. We'll assume the former for
```

Comment:
Why assume []MVCCKeyValue for the "effects" contents? Then we're tying up potential for migration in how we encode MVCC information. I think it's more prudent to send "post-MVCC raw" key-value pairs.